### PR TITLE
Make constraint error details contain useful information for developers

### LINKED ISSUE #6768: `message` and `details` fields are the same when you specify an `errmessage` for a constraint.
- EdgeDB Version: 4.4+aa576e1
- EdgeDB CLI Version:  4.0.2+500be79
- OS Version: macOS ventura

Steps to Reproduce:

1. For one of the properties in your schema add a `errmessage`, to customize the default errors emiitted for constraint validation.
2. Try to insert a record using the edgedb repl that fails this constraint
3. It returns the same error for `message` as well as `detail`

Screenshot 2024-02-01 at 17 34 03

Schema:
```
  type User extending Model {
    first_name: str;
    last_name: str;
    required email: str {
      constraint exclusive {
        errmessage := 'non-unique email'
      };
      constraint regexp(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$') {
        errmessage := 'invalid email, try again later'
      };
    };
    is_email_verified: bool {
      default := false;
    };
  }
```

Originally asked here:
https://discord.com/channels/841451783728529451/849374705210490900/1202341124673634304

### PR DESCRIPTION
**Pull request by aljazerzen**
Closes #6768

Changes error details of constraint violations to:

```
f'violated {constraint_description} on {subject_description}'
```
... where `x_description` is the `x.get_verbosename()`

### CODE REVIEW

File: edb/schema/constraints.py
Comment (aljazerzen) [inline edb/schema/constraints.py:199]: This is a refactor, me trying to understand if I can use this code for my case

File: edb/server/compiler/errormech.py
Comment (aljazerzen) [inline edb/server/compiler/errormech.py:631]: This was my initial approach for the details

File: tests/test_constraints.py
Comment (aljazerzen) [inline tests/test_constraints.py:1162]: Am I doing something wrong here? Or does edgedb-python really don't expose error details field?

Comment (aljazerzen) [reply tests/test_constraints.py:1162]: Turns out that edgedb-python really does not expose details.

File: edb/server/compiler/errormech.py
Comment (elprans) [inline edb/server/compiler/errormech.py:602]:
```suggestion
```

File: edb/schema/constraints.py
Comment (msullivan) [reply edb/schema/constraints.py:199]: Probably not, right?

File: edb/schema/constraints.py
Comment (msullivan) [inline edb/schema/constraints.py:185]: I don't think this explanation is quite right. I think it would be worth having clear documentation here, because there is a subtle issue at play.

The purpose of this function is to find which types an exclusive constraint originate from. If we have `Baz <: Bar <: Foo`, and Foo declares some exclusive constraint, then when we do an insert into `Baz` we need to check all `Foo`s to make sure there is no conflict. (Which we do using triggers, and `get_constraint_origin` helps drive the generation of those.)

The subtleties here are:
 1. Because of multiple inheritance, we might have an exclusive constraint that originates from multiple unrelated ancestors
 2. Delegated exclusive constraints are declared on a parent, but their "origins" are their children. (So that `abstract type Named` or whatever can declare an exclusive name, but the names will be exclusive *within* each of the types that subtypes it.)

**Review by msullivan — APPROVED** [review]

File: edb/schema/constraints.py
Comment (aljazerzen) [reply edb/schema/constraints.py:199]: No, it does something else (but that was not obvious to me before the refactor)

File: tests/test_constraints.py
Comment (aljazerzen) [reply tests/test_constraints.py:1162]: I'll wait for a release of edgedb-python and clean this up before merging.

File: edb/schema/constraints.py
Comment (aljazerzen) [reply edb/schema/constraints.py:185]: I'll add these docs in a separate PR

**Comment by msullivan** [comment]
Let's merge it without waiting for edgedb-python, so we can backport it without needing to bump edgedb-python on those branches

**Pull request merged**
